mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala

```diff
@@ -474,6 +487,75 @@ class LinearRegressionSummary private[regression] (
     predictions.select(t(col(predictionCol), col(labelCol)).as("residuals"))
   }
 
+  /** Number of instances in DataFrame predictions */
+  lazy val numInstances: Long = predictions.count()
+
+  /** Degrees of freedom */
+  private val degreesOfFreedom: Long = if (model.getFitIntercept) {
+    numInstances - model.coefficients.size - 1
+  } else {
+    numInstances - model.coefficients.size
+  }
+
+  /**
+   * The weighted residuals, the usual residuals rescaled by
+   * the square root of the instance weights.
+   */
+  lazy val devianceResiduals: Array[Double] = {
```

Comment:
I'm late to comment, but am wondering:
* Why do we not return all deviance residuals as a DataFrame?  If we only return min,max, then that should be documented.  But I'd prefer we return a DataFrame with all deviance residuals.
* Should we follow R's example and just call this "residuals"?  That will let us add other types of residuals later (specified via an argument, with a default argument of "deviance").